### PR TITLE
feat: add semantic-release to d2-app scripts

### PR DESCRIPTION
**Pull request by varl**
This adds the `release` command under `d2-app scripts`. It is based on `semantic-release` so it is fully automated. Right now it doesn't allow overrides to its configuration from the application/library and uses "sane defaults" (i.e. the way we want it to behave). When it's proven to work we can add the ability to merge configurations from the application/library.

### CODE REVIEW

File: packages/d2-app/src/commands/scripts/release.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        ],
+
+        '@semantic-release/npm',
+
+        [
+            '@semantic-release/git',
+            {
+                assets: ['CHANGELOG.md', 'package.json', 'yarn.lock'],
+                message:
+                    'chore(release): cut ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}',
```
Comment (varl) [inline packages/d2-app/src/commands/scripts/release.js]: https://changelog.travis-ci.com/skip-travis-ci-builds-with-skip-travis-in-commit-message-74620

Not sure if we want this to skip CI. :thinking:

File: packages/d2-app/package.json [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
-    "description": "D2 CLI module for interacting with frontend applications",
-    "main": "index.js",
-    "repository": "https://github.com/dhis2/cli",
-    "author": "Austin McGee <austin@dhis2.org>",
+    "bin": {
+        "d2-app": "./bin/d2-app"
+    },
+    "version": "1.0.0",
+    "main": "src/index.js",
+    "author": "DHIS2 &lt;info@dhis2.org&gt;",
```
Comment (amcgee) [inline packages/d2-app/package.json]: This is a bug in the `create` template, need to un-escape it somehow

Comment (amcgee) [reply packages/d2-app/package.json]: But for now maybe manually is fine?

File: packages/d2-app/src/commands/scripts/release.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            ...process.env,
+            GIT_AUTHOR_NAME: '@dhis2-bot',
+            GIT_AUTHOR_EMAIL: 'ci@dhis2.org',
+            GIT_COMMITTER_NAME: '@dhis2-bot',
+            GIT_COMMITTER_EMAIL: 'ci@dhis2.org',
+        },
+    }
+
+    try {
+        const result = await semanticRelease(options, config)
```
Comment (amcgee) [inline packages/d2-app/src/commands/scripts/release.js]: We should add a `--dry-run` CLI option which passes through to the same option for semantic release, so we can test release logic locally

File: packages/d2-app/src/commands/scripts/release.js (hunk 1 above)
Comment (amcgee) [reply packages/d2-app/src/commands/scripts/release.js]: Hmm yeah probably fine for now?

**Comment by amcgee** [comment]
Really cool to get this all automated!

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: I can do that. Right now you are barred from doing a local release as it checks for a CI environment by default. I'll make that explicit too.

**Comment by amcgee** [comment]
Also also - how do we do this in mono-repos?  Would love to use it in `dhis2/cli` itself, but I think we'd need to restructure the repo, use [this](https://github.com/Updater/semantic-release-monorepo) which doesn't look super solid, or make our own set of configs for mono-repo support (and particularly single-version monorepo support?)

File: packages/d2-app/package.json (hunk 2 above)
Comment (varl) [reply packages/d2-app/package.json]: 86c2edd

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: Introduced `--no-dry-run`/`--dry-run` and `--no-ci`/`--ci`

Comment (amcgee) [reply packages/d2-app/src/commands/scripts/release.js]: Do you think we could get away with not allowing `--ci`?  Maybe we need it

**Comment by varl** [comment]
> Also also - how do we do this in mono-repos? Would love to use it in `dhis2/cli` itself and with `d2-ui` for instance, but I think we'd need to restructure the repo, use [this](https://github.com/Updater/semantic-release-monorepo) which doesn't look super solid, or make our own set of configs for mono-repo support (and particularly single-version monorepo support?)

Yeah, I thought I read about this working, but I was mistaken: https://github.com/semantic-release/semantic-release/issues/193

This is likely more solid, but it uses commitizen: https://github.com/atlassian/lerna-semantic-release

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: We can get away with not having `--dry-run` and `--ci`, I think.

`--ci` defaults to `true` in semantic-release, so the default is to only allow CI environments to run the release.

So if I run the release command just the defaults it automatically goes into dry run mode.

```
d2 app scripts release
...
[1:59:08 PM] [semantic-release] › ⚠  This run was not triggered in a known CI environment, running in dry-run mode.
```

To actually run the `release` locally I need to pass in `--no-ci`.

**Comment by amcgee** [comment]
Oh cool!  I thought we had to pass --dry-run but if it skips when not in a
CI that's even better.  Sorry for the misdirection

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: >Also, is this actually for apps or just for libraries? Apps still have a different (non-npm) release process right?

Yeah, I'm pondering adding a `--mode [app/lib]` switch, as the only thing we would need to _not_ do for apps is publish to NPM:  https://github.com/dhis2/cli/blob/86c2edd83d92d72f08115955c0e0b377c2591d4c/packages/d2-app/src/commands/scripts/release.js#L24

**Comment by amcgee** [comment]
What about a `--publish` flag?

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (amcgee) [reply packages/d2-app/src/commands/scripts/release.js]: What about a `--publish` flag, set for libs but not apps?  Or check the `publishConfig` in package.json?

( GitHub hates email )

Comment (amcgee) [reply packages/d2-app/src/commands/scripts/release.js]: > So if I run the release command with the defaults it automatically goes into dry run mode.

Cool, yeah no need for dry-run then!  Sorry I didn't realize you could preview it locally without

Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: Sure. :+1:

An alternative is just setting `"private": true"` in `package.json` for Apps (not verified).

Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: I would like to have a `--publish [target]` in that case, so we can publish to NPM for libs and App Store for apps.

Comment (varl) [reply packages/d2-app/src/commands/scripts/release.js]: 0c6fd99

**Comment by varl** [comment]
Review notes:

- dryRun by default on local machines
- Do not publish by default
- Allow for publish providers through `--publish [provider]`
- Only provider now is NPM

File: packages/d2-app/package.json [hunk 4]
```diff
@@ -1,18 +1,22 @@
...
+    "bin": {
+        "d2-app": "./bin/d2-app"
+    },
     "version": "0.11.0",
-    "description": "D2 CLI module for interacting with frontend applications",
-    "main": "index.js",
-    "repository": "https://github.com/dhis2/cli",
-    "author": "Austin McGee <austin@dhis2.org>",
+    "main": "src/index.js",
+    "author": "Viktor Varland <viktor@dhis2.org>",
```
Comment (amcgee) [inline packages/d2-app/package.json]: I see what you did there ;-)

Comment (varl) [reply packages/d2-app/package.json]: Hahaha, oh man, sorry about that.

I did a `rm` of `d2-app`, then regenerated it with the `create cli` tool to allow it to match the new convention as it seemed to predate the create script, and accepted the defaults. :D

File: packages/d2/src/index.js [hunk 5]
```diff
@@ -4,11 +4,12 @@ const command = namespace('d2', {
     desc: 'DHIS2 CLI',
     builder: yargs => {
         yargs.commandDir('commands')
-        yargs.command(require('@dhis2/cli-create'))
-        yargs.command(require('@dhis2/cli-app'))
         yargs.command(require('@dhis2/cli-cluster'))
+        yargs.command(require('@dhis2/cli-packages'))
```
Comment (amcgee) [inline packages/d2/src/index.js]: Get outa here!

File: packages/d2-app/src/commands/scripts/release.js (hunk 3 above)
Comment (amcgee) [reply packages/d2-app/src/commands/scripts/release.js]: This solution looks good to me!

File: packages/d2-app/package.json (hunk 4 above)
Comment (varl) [reply packages/d2-app/package.json]: 8352ec4

File: packages/d2/src/index.js (hunk 5 above)
Comment (varl) [reply packages/d2/src/index.js]: What the!

e4e4f53

**Pull request merged**